File: source/yeon/7주차/code/1074Z.cpp
// 백준 1074 Z 
// cpp solved by divide and conquer
#include<iostream>
int N, r, c;
long long ans = 0;
void solution(int y, int x, int n){
    if (y == r && x == c) {
        printf("%lld\n", ans);
        return;
    }

    // find the location of r,c 
    // if they are in boundary divide 
    // not in booundary just add n*n 
    if (c>= x && c< x+n && r>=y && r< y+n){
        solution(y,x,n/2);
        solution(y,x+n/2,n/2);
        solution(y+n/2,x,n/2);
        solution(y+n/2,x+n/2,n/2);
    }else{
        ans += n * n;
    }
}
int main(void){
    scanf("%d %d %d", &N, &r, &c);
    // 1<<N => 2^N
    solution(0,0,1<<N);
    return 0;
}

File: source/yeon/7주차/code/1074Z.py
# 백준 1074번 Z 
# python solved by divide and conquer
N, r, c = map(int,input().split())
ans = 0

def solution(y:int, x:int, n: int):
    global ans
    if y == r and x == c : 
        print(ans) 
        return 
    
    # find the location of r,c 
    # if they are in boundary divide 
    # not in booundary just add n*n 
    if y <= r < y + n and x <= c < x + n :
        solution(y, x, n // 2)
        solution(y, x+n//2, n//2)
        solution(y+n//2, x, n//2)
        solution(y+n//2, x+n//2, n//2)
    else :
        ans += n * n 
        
solution(0,0,1<<N)
